langchain/callbacks/tracers/base.py

```diff
@@ -58,235 +68,211 @@ def load_session(self, session_name: str) -> TracerSession:
+
+    def _end_trace(self, run) -> None:
         """End a trace for a run."""
-        run = self._stack.pop()
-        if not self._stack:
-            self._execution_order = 1
+        if not run.parent_uuid:
             self._persist_run(run)
+            self.execution_order = 1
+        self.run_map.pop(run.uuid)
 
     def on_llm_start(
-        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
+        self,
+        serialized: Dict[str, Any],
+        prompts: List[str],
+        run_id: str = None,
+        parent_run_id: str = None,
+        **kwargs: Any,
     ) -> None:
         """Start a trace for an LLM run."""
-        if self._session is None:
-            raise TracerException(
-                "Initialize a session with `new_session()` before starting a trace."
-            )
+        if self.session is None:
+            self.session = self.load_default_session()
+
+        if run_id is None:
+            run_id = str(uuid4())
 
         llm_run = LLMRun(
+            uuid=run_id,
+            parent_uuid=parent_run_id,
             serialized=serialized,
             prompts=prompts,
             extra=kwargs,
             start_time=datetime.utcnow(),
-            execution_order=self._execution_order,
-            session_id=self._session.id,
-            id=self._generate_id(),
+            execution_order=self.execution_order,
+            session_id=self.session.id,
         )
         self._start_trace(llm_run)
 
-    def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
-        """Handle a new token for an LLM run."""
-        pass
-
-    def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
+    def on_llm_end(
+        self, response: LLMResult, run_id: str = None, **kwargs: Any
+    ) -> None:
         """End a trace for an LLM run."""
-        if not self._stack or not isinstance(self._stack[-1], LLMRun):
-            raise TracerException("No LLMRun found to be traced")
+        if not run_id:
+            raise TracerException("No run_id provided for on_llm_end callback.")
 
-        self._stack[-1].end_time = datetime.utcnow()
-        self._stack[-1].response = response
+        if not self.run_map or not isinstance(self.run_map[run_id], LLMRun):
+            raise TracerException("No LLMRun found to be traced")
 
-        self._end_trace()
+        llm_run = self.run_map[run_id]
+        llm_run.response = response
+        llm_run.end_time = datetime.utcnow()
+        self._end_trace(llm_run)
 
     def on_llm_error(
-        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+        self,
+        error: Union[Exception, KeyboardInterrupt],
+        run_id: str = None,
+        **kwargs: Any,
     ) -> None:
         """Handle an error for an LLM run."""
-        if not self._stack or not isinstance(self._stack[-1], LLMRun):
-            raise TracerException("No LLMRun found to be traced")
+        if not run_id:
```

Comment:
We need to make this a required arg in all methods (other than handle_start_* in CallbackManager)